File: Exams/Resoluções de Exames by Carlos Martins/examples/Example2.java
/**
 *  ISEL, LEIC, Concurrent Programming
 *
 *  Unfair semaphore implementation using the pattern presented in the
 *  Example 2 of "Sincronização com Monitores na CLI e na Infra-estrutura Java".
 *
 *  Generate executable with: javac Example2.java TimeoutHolder.java
 *  Execute with: java Example2
 *
 *  Carlos Martins, April 2018
 *
 **/

import java.util.Random;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.atomic.AtomicInteger;

final class Semaphore_ {
	private final Object monitor = new Object();
	private int permits;
	
	// Constructors
	public Semaphore_(int initial) {
		if (initial > 0)
			permits = initial;
	}
	
	public Semaphore_() {}

	// Acquire "acquires" permits with timeout
	public boolean acquire(int acquires, long timeout, TimeUnit unit) throws InterruptedException {
		synchronized(monitor) {
			// if there are sufficient permits, update semaphore value and return success
			if (permits >= acquires) {
				permits -= acquires;
				return true;
			}
			
			// test if it is about immediate acquisition and return failure if it is the case
			if (timeout == 0)
				return false;

			// create an instance of TimeoutHolder to support timeout adjustments
			TimeoutHolder th = new TimeoutHolder(timeout, unit);
			
			// loop until acquire n permits, the specified timeout expires or thread
			// is interrupted
			do {
				if (th.isTimed()) {
					if ((timeout = th.value()) == 0)
						return false;			// timeout expired
					monitor.wait(timeout);
				} else
					monitor.wait();
			} while	(permits < acquires);
			// permits available
			permits -= acquires;
			return true;
		}
	}
	
	// Acquire "acquires" permits with a millis timeout 
	public boolean acquire(int acquires, long millisTimeout) throws InterruptedException {
		return acquire(acquires, millisTimeout, TimeUnit.MILLISECONDS);
	}
	
	// Acquire "acquires" permits without timeout 
	public void acquire(int acquires) throws InterruptedException {
		acquire(acquires, -1L, TimeUnit.MILLISECONDS);
	}
	
	// Release "releases" permits
	public void release(int releases){
		synchronized(monitor) {
			permits += releases;
			monitor.notifyAll();
		}
	}
}

public class Example2 {
	
	private static boolean testUnfairness() throws InterruptedException {

		final int RUN_TIME = 10 * 1000;
		final int SETUP_TIME = 20;
		final int THREADS = 50;
		final int MIN_TIMEOUT = 1;
		final int MAX_TIMEOUT = 10;
		

		Thread[] tthrs = new Thread[THREADS];
		int[] privateCounters = new int[THREADS];
		CountDownLatch start = new CountDownLatch(1);
		Semaphore_ sem = new Semaphore_(THREADS);
		AtomicInteger totalTimeouts = new AtomicInteger(0);
		
		System.out.println("-->test semaphore unfairness");
		for (int i = 0; i < THREADS; i++) {
			final int tid = i;
			tthrs[i] = new Thread(() -> {
				Random random = new Random(tid);			
				long deadline = System.currentTimeMillis() + RUN_TIME;
				try {
					// Wait until start event is set
					start.await();
					do {
						do {
							if (sem.acquire(tid + 1, random.nextInt(MAX_TIMEOUT) + MIN_TIMEOUT))
								break;
							totalTimeouts.incrementAndGet();
						} while (true);
						Thread.yield();
						sem.release(tid + 1);
						if ((++privateCounters[tid] % 1000) == 0)
							System.out.printf("[#%02d]", tid);
					} while (System.currentTimeMillis() < deadline);
				} catch (InterruptedException ie) {
					// this never happens
				}				
			});
			tthrs[i].start();
		}
		
		// Wait until all test threads have been started and then set the
		// start event.
		Thread.sleep(SETUP_TIME);
		start.countDown();
		
		// Wait until all threads have been terminated.
		for (int i = 0; i < THREADS; i++)
			tthrs[i].join();
		
		// Show results
		int total = 0;
		System.out.println("\n\nPrivate counters:");
		for (int i = 0; i < THREADS; i++) {
			if ((i % 5) == 0)
				System.out.println();
			else if (i != 0)
				System.out.print(' ');
			System.out.printf("[#%02d:%5d]", i, privateCounters[i]);
			total += privateCounters[i];
		}
		System.out.printf("%n--total acquisitions/releases: %d, timeouts: %d%n", total, totalTimeouts.get());
		return true;
	}


	public static void main(String... args) throws InterruptedException {
		System.out.printf("%n-->Test semaphore unfairness: %s%n",
						  testUnfairness() ? "passed" : "failed");
	}
}

File: Exams/Resoluções de Exames by Carlos Martins/examples/Example2.cs
/**
 *  ISEL, LEIC, Concurrent Programming
 *
 *  Unfair semaphore implementation using the pattern presented in the
 *  Example 2 in "Sincronização com Monitores na CLI e na Infra-estrutura Java".
 *
 *  Generate executable with: csc Example2.cs TimeoutHolder.cs
 *
 *  Carlos Martins, April 2018
 *
 **/

using System;
using System.Threading;

internal sealed class Semaphore_ {
	private readonly object monitor = new object();
	private int permits;
	
	// Constructor
	public Semaphore_ (int initial = 0) {
		if (initial > 0)
			permits = initial;
	}
	
	// Acquire "acquires" permits
	public bool Acquire(int acquires, int timeout = Timeout.Infinite) {
		lock(monitor) {
			// if there are sufficient permits, update semaphore value and return success
			if (permits >= acquires) {
				permits -= acquires;
				return true;
			}
			
			// test if it is about immediate acquisition and return failure if it is the case
			if (timeout == 0)
				return false;

			// create an instance of TimeoutHolder to support timeout adjustments
			TimeoutHolder th = new TimeoutHolder(timeout);
			
			// loop until acquire n permits, the specified timeout expires or thread
			// is interrupted
			do {
				if ((timeout = th.Value) == 0)
					return false;			// timeout expired
				Monitor.Wait(monitor, timeout);
			} while	(permits < acquires);
			// permits available
			permits -= acquires;
			return true;
		}
	}
	
	// Release "releases" permits
	public void Release(int releases){
		lock(monitor) {
			permits += releases;
			Monitor.PulseAll(monitor);
		}
	}
}

public class Example2 {
	
	private static bool TestUnfairness() {

		const int RUN_TIME = 10 * 1000;
		const int SETUP_TIME = 20;
		const int THREADS = 50;
		const int MIN_TIMEOUT = 1;
		const int MAX_TIMEOUT = 10;
		

		Thread[] tthrs = new Thread[THREADS];
		int[] privateCounters = new int[THREADS];
		ManualResetEventSlim startEvent = new ManualResetEventSlim(false);
		Semaphore_ sem = new Semaphore_(THREADS);
		int totalTimeouts = 0;
		
		Console.WriteLine("-->test semaphore unfairness");
		for (int i = 0; i < THREADS; i++) {
			int tid = i;
			tthrs[i] = new Thread(() => {
				Random random = new Random(tid);
			
				// Wait until start event is set
				startEvent.Wait();
			
				int endTime = Environment.TickCount + RUN_TIME;
				
				do {
					do {
						if (sem.Acquire(tid + 1, random.Next(MIN_TIMEOUT, MAX_TIMEOUT)))
							break;
						Interlocked.Increment(ref totalTimeouts);
					} while (true);
					Thread.Yield();
					sem.Release(tid + 1);
					if ((++privateCounters[tid] % 200) == 0) {
						Console.Write($"[#{tid:D2}]");
					}
				} while (Environment.TickCount < endTime);					
			});
			tthrs[i].Start();
		}
		
		// Wait until all test threads have been started and then set the
		// start event.
		Thread.Sleep(SETUP_TIME);
		startEvent.Set();
		
		// Wait until all threads have been terminated.
		for (int i = 0; i < THREADS; i++)
			tthrs[i].Join();
		
		// Show results
		int total = 0;
		Console.WriteLine("\n\nPrivate counters:");
		for (int i = 0; i < THREADS; i++) {
			if (i != 0 && (i % 5) == 0)
				Console.WriteLine();
			else if (i != 0)
				Console.Write(' ');
			Console.Write($"[#{i:D2}:{privateCounters[i],4}]");
			total += privateCounters[i];
		}
		Console.WriteLine($"\n--total acquisitions/releases: total, timeouts: {totalTimeouts}");
		return true;
	}


	public static void Main() {
		Console.WriteLine("\n-->Test semaphore unfairness: {0}",
						  TestUnfairness() ? "passed" : "failed");
	}
}
